### PR TITLE
SCP-2190 2-Phase validation

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract/Wallet.hs [hunk 1]
```diff
@@ -134,7 +134,13 @@ balanceTx ::
...
     -> Eff effs Tx
 balanceTx utxo pk UnbalancedTx{unBalancedTxTx} = do
-    (neg, pos) <- Value.split <$> computeBalance unBalancedTxTx
+    (neg', pos) <- Value.split <$> computeBalance unBalancedTxTx
+
+    -- Make sure fees are always added to `neg` instead of subtracted from `pos`.
+    -- This way inputs will always be created for fees, which are added to the txInputsFees field.
+    -- This is not the optimal strategy, but it does ensure that at least all the fees will be
+    -- covered if the transaction turns out to be invalid.
+    let neg = neg' P.+ L.txFee unBalancedTxTx
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/Wallet.hs:143]: Doesn't `computeBalance` include the fees already?

Comment (j-mueller) [reply plutus-contract/src/Plutus/Contract/Wallet.hs:143]: If always add the fee to `neg` then `balanaceTx` isn't idempotent anymore

File: plutus-contract/src/Plutus/Contract/Wallet.hs [hunk 2]
```diff
@@ -168,7 +174,7 @@ addInputs mp pk vl tx = do
 
         addTxIns  =
             let ins = Set.fromList (Tx.pubKeyTxIn . fst <$> spend)
-            in over Tx.inputs (Set.union ins)
+            in over Tx.inputsFees (Set.union ins)
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/Wallet.hs:177]: Sometimes the tx inputs that are added here have nothing to do with the fees. For example if just want to include a token in the transaction. In that case the input should be added to `Tx.inputs`. Maybe we should first balance the transaction, putting all new inputs into `Tx.inputs`, and then in a second step move some of the inputs from `Tx.inputs` to `Tx.inputsFees`. Then we also won't need to add the fees to `neg'` (I think)

File: plutus-ledger/src/Ledger/Index.hs [hunk 3]
```diff
@@ -151,21 +161,30 @@ lkpTxOut t = lookup t =<< ask
...
-    emptyUtxoSet <- reader (Map.null . getIndex)
-    unless emptyUtxoSet (checkForgingScripts t)
-    unless emptyUtxoSet (checkForgingAuthorised t)
-    unless emptyUtxoSet (checkTransactionFee t)
-
-    _ <- checkValidInputs t
-    insert t <$> ask
+    _ <- checkValidInputs (view inputsFees) t
+    (do
+        _ <- checkSlotRange h t
```
Comment (j-mueller) [inline plutus-ledger/src/Ledger/Index.hs:168]: Phase 2 is only for script validation, everything else (value preserved, positive value, witnesses for public key inputs, slot range) needs to be validated in phase 1.

Also I think we shoud check that `inputsFees`  and `inputs` don't overlap.

File: plutus-playground-server/test/Playground/UsecasesSpec.hs [hunk 4]
```diff
@@ -174,12 +175,12 @@ gameTest =
     testGroup
         "game"
         [ compilationChecks game
-        , testCase "should keep the funds" $
-          evaluate (mkEvaluation [lock w2 "abcde" twoAda, AddBlocks 1, guess w1 "ade", AddBlocks 1]) >>=
-          hasFundsDistribution
-              [ mkSimulatorWallet w1 tenAda
-              , mkSimulatorWallet w2 (adaValueOf 8)
-              ]
+        -- , testCase "should keep the funds" $
```
Comment (j-mueller) [inline plutus-playground-server/test/Playground/UsecasesSpec.hs:178]: Are these the tests that will be enabled again later on?

Comment (sjoerdvisscher) [reply plutus-playground-server/test/Playground/UsecasesSpec.hs:178]: Yes

File: plutus-ledger/src/Ledger/Index.hs [hunk 5]
```diff
@@ -163,20 +165,22 @@ validateTransaction :: ValidationMonad m
...
-        emptyUtxoSet <- reader (Map.null . getIndex)
-        unless emptyUtxoSet (checkForgingScripts t)
-        unless emptyUtxoSet (checkForgingAuthorised t)
-        unless emptyUtxoSet (checkTransactionFee t)
+    -- see note [Forging of Ada]
+    emptyUtxoSet <- reader (Map.null . getIndex)
+    unless emptyUtxoSet (checkForgingAuthorised t)
+    unless emptyUtxoSet (checkTransactionFee t)
+
+    checkValidInputs (view inputs) t
```
Comment (j-mueller) [inline plutus-ledger/src/Ledger/Index.hs:178]: I think we need to split up `checkValidInputs` into `checkValidPubKeyInputs` and `checkValidScriptInputs` - in phase 2 the script inputs should be checked

**Comment by sjoerdvisscher** [comment]
@j-mueller Does the new balancing code make sense to you? It turned out to be pretty complicated.

**Review by j-mueller — APPROVED** [review]
Great! Let's merge it!

**Pull request merged**